File: src/main/java/org/kite9/diagram/visualization/planarization/mgt/MGTVertexEdgeOrdering.java
package org.kite9.diagram.visualization.planarization.mgt;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.edge.PlanarizationEdge;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.planarization.Tools;
import org.kite9.diagram.visualization.planarization.ordering.AbstractCachingEdgeOrdering;
import org.kite9.diagram.visualization.planarization.ordering.VertexEdgeOrdering;

public class MGTVertexEdgeOrdering extends AbstractCachingEdgeOrdering implements VertexEdgeOrdering {
	
	MGTPlanarization pl;

	Vertex v;
	
	List<PlanarizationEdge> af;
	List<PlanarizationEdge> ab;
	List<PlanarizationEdge> bf;
	List<PlanarizationEdge> bb;
	
	public MGTVertexEdgeOrdering(MGTPlanarization pl, Vertex v) {
		this.pl = pl;
		this.v = v;
		this.af = pl.getAboveForwardLinks(v);
		this.bf = pl.getBelowForwardLinks(v);
		this.bb = pl.getBelowBackwardLinks(v);
		this.ab = pl.getAboveBackwardLinks(v);
	}

	
	@Override
	public void remove(PlanarizationEdge toRemove) {
		safeRemove(toRemove, af);
		safeRemove(toRemove, bf);
		safeRemove(toRemove, bb);
		safeRemove(toRemove, ab);
		changed();
	}
	
	private void safeRemove(PlanarizationEdge toRemove, List<PlanarizationEdge> l) {
		if (l!=null) {
			l.remove(toRemove);
		}
	}

	
	@Override
	public int size() {
		return af.size()+bf.size()+bb.size()+ab.size();
	}
	
	@Override
	protected List<PlanarizationEdge> getEdgesAsListInner() {
		List<PlanarizationEdge> out = new ArrayList<PlanarizationEdge>(size());
		addAllBackwards(out, af);
		out.addAll(bf);
		addAllBackwards(out, bb);
		out.addAll(ab);
		
		directions = null;
		for (Edge edge : out) {
			addEdgeDirection(edge.getDrawDirectionFrom(v), Tools.isUnderlyingContradicting(edge));
		}
		
		return Collections.unmodifiableList(out);
	}

	private void addAllBackwards(List<PlanarizationEdge> out, List<PlanarizationEdge> af2) {
		for (int i = af2.size()-1; i >=0; i--) {
			out.add(af2.get(i));
		}
	}

	@Override
	public String toString() {
		return "[VEO:"+(getEdgeDirections()==MUTLIPLE_DIRECTIONS ? "MULTI" : getEdgeDirections())+":"+getEdgesAsList()+"]";
	}


	@Override
	public void replace(PlanarizationEdge b, PlanarizationEdge a) {
		replace(af, b, a);
		replace(bf, b, a);
		replace(bb, b, a);
		replace(ab, b, a);
		changed();
	}
	
	private boolean replace(List<PlanarizationEdge> in, PlanarizationEdge b, PlanarizationEdge a) {
		int bIndex = in.indexOf(b);
		if (bIndex!=-1) {
			in.set(bIndex, a);
			return true;
		} else {
			return false;
		}
	}

	@Override
	protected Direction getInterceptDirection(Edge e) {
		return e.getDrawDirectionFrom(v);
	}

	
}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/mgt/MGTVertexEdgeOrdering.kt
package org.kite9.diagram.visualization.planarization.mgt

import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.visualization.planarization.Tools.Companion.isUnderlyingContradicting
import org.kite9.diagram.visualization.planarization.ordering.AbstractCachingEdgeOrdering
import org.kite9.diagram.visualization.planarization.ordering.VertexEdgeOrdering
import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.model.position.Direction
import java.util.Collections
import org.kite9.diagram.visualization.planarization.ordering.EdgeOrdering
import java.util.ArrayList

class MGTVertexEdgeOrdering(val pl: MGTPlanarization, val v: Vertex) : AbstractCachingEdgeOrdering(),
    VertexEdgeOrdering {

    var af: MutableList<PlanarizationEdge> = pl.getAboveForwardLinks(v)
    var ab: MutableList<PlanarizationEdge> = pl.getAboveBackwardLinks(v)
    var bf: MutableList<PlanarizationEdge> = pl.getBelowForwardLinks(v)
    var bb: MutableList<PlanarizationEdge> = pl.getBelowBackwardLinks(v)

    override fun remove(toRemove: PlanarizationEdge) {
        safeRemove(toRemove, af)
        safeRemove(toRemove, bf)
        safeRemove(toRemove, bb)
        safeRemove(toRemove, ab)
        changed()
    }

    private fun safeRemove(toRemove: PlanarizationEdge, l: MutableList<PlanarizationEdge>) {
        l.remove(toRemove)
    }

    override fun size(): Int {
        return af.size + bf.size + bb.size + ab.size
    }

    override val edgesAsListInner: List<PlanarizationEdge>
        protected get() {
            val out: MutableList<PlanarizationEdge> = ArrayList(size())
            addAllBackwards(out, af)
            out.addAll(bf)
            addAllBackwards(out, bb)
            out.addAll(ab)
            directions = null
            for (edge in out) {
                addEdgeDirection(edge.getDrawDirectionFrom(v), isUnderlyingContradicting(edge))
            }
            return Collections.unmodifiableList(out)
        }

    private fun addAllBackwards(out: MutableList<PlanarizationEdge>, af2: List<PlanarizationEdge>) {
        for (i in af2.indices.reversed()) {
            out.add(af2[i])
        }
    }

    override fun toString(): String {
        return "[VEO:" + (if (getEdgeDirections() === EdgeOrdering.MUTLIPLE_DIRECTIONS) "MULTI" else getEdgeDirections()) + ":" + getEdgesAsList() + "]"
    }

    override fun replace(b: PlanarizationEdge, a: PlanarizationEdge) {
        replace(af, b, a)
        replace(bf, b, a)
        replace(bb, b, a)
        replace(ab, b, a)
        changed()
    }

    private fun replace(ml: MutableList<PlanarizationEdge>, b: PlanarizationEdge, a: PlanarizationEdge): Boolean {
        val bIndex = ml.indexOf(b)
        return if (bIndex != -1) {
            ml.set(bIndex, a)
            true
        } else {
            false
        }
    }

    override fun getInterceptDirection(e: Edge): Direction {
        return e.getDrawDirectionFrom(v)!!
    }

}